### ISSUE TITLE
Unable to complete MFA sign in flow using `startActivityForSignInWithProvider`

### ISSUE THREAD

**Issue by samrit** [firebase/firebase-android-sdk#2741]
### Environment
1. com.google.firebase:firebase-auth:21.0.1
2. com.google.android.gms:play-services-auth:19.0.0
3. Android 11
4. Google Identity platform
5. Custom OIDC Provider (any)

### The problem

We are integrating with our customers existing identity stores using the Google Cloud Identity platform. Some of our customers require MFA using various MFA apps. We are using the Android SDK method `startActivityForSignInWithProvider` ([described here](https://developers.google.com/android/reference/com/google/firebase/auth/FirebaseAuth#startActivityForSignInWithProvider(android.app.Activity,%20com.google.firebase.auth.FederatedAuthProvider))). Now this works great on iOS and in the browser. But as implemented currently, it is impossible for a user to complete a MFA sign in flow using a custom provider, on Android, that utilizes a MFA app that is installed on the same device.

This is because the browser window opened by the SDK to complete the OAuth flow closes, as soon as a user tries to switch tasks. So any attempt to switch to the MFA app, either to copy a code or to approve the sign in request, kills the sign in flow. See attached video for a demonstration of the problem (https://user-images.githubusercontent.com/4655446/122409397-19f94800-cf51-11eb-9c23-4843de51fbf6.mp4).

Our hypothesis is that the flags used to open the browser window are wrong. See https://github.com/openid/AppAuth-Android/issues/106 and https://github.com/openid/AppAuth-Android/pull/109.

#### Steps to reproduce:
1. Configure a custom identity provider using Google Identity Provider (SAML or OIDC). 
2. Make sure your custom third party identity server enforces MFA using some authenticator app
3. Start login flow using `startActivityForSignInWithProvider`
4. Switch tasks

### Critical
This is slowly becoming a critical issue as many of our enterprise customers are reporting issues for their employees that use Android. Given that the fix is likely easy, hoping for a quick resolution

**Comment by aguatno** [comment]
Hi @samrit thanks for reporting this. 

Hi @malcolmdeck this one seems unusual. Any thoughts why the OAuth browser closes when a user tries to switch task? I found this [issue](https://github.com/firebase/firebase-android-sdk/issues/478) that could be related to this and it looks like we need a bug report?

**Comment by rosalyntan** [comment]
(Googler-only internal tracking bug: b/191682367)

**Comment by samrit** [comment]
Hey all, I know you guys are on this but we just lost a customer because of this issue. Is there anyway we can get a time estimate? It will help us manage expectations.

**Comment by samrit** [comment]
@rosalyntan Is there a workaround we could use. This is a pretty major issue for us.

**Comment by deskbirdbob** [comment]
I am also experiencing this exact issue with our android app. We have a lot of customers signing in via Microsoft and the sign in flow cancels itself whenever they try to switch to their MFA authenticator app.

The exception I am getting in logs is this:
`com.google.firebase.auth.FirebaseAuthWebException: The web operation was canceled by the user.`
Note that I am not actually canceling anything, the chrome tab disappears immediately after it loses focus and returns this error.

Any news about this issue? Is there something I could do to fix it?

**Comment by fabricio-godoi** [comment]
I found a user-side workaround for now. Once the login process starts, request the Custom Tab to open in the actual browser instead of following the process in the Custom Tab.

Because the Custom Tab is attached to the main app, and the main app is put to rest once it is not the main app, I think that some Custom Tab configuration is missing to prevent its closure. If you redirect to the actual browser, the issue does not happen.

I don´t think this is a valid solution for the problem, but if your user base is not that big, this could help using some MFA with this method.

![image](https://user-images.githubusercontent.com/7767910/134964746-9c93c1c2-e291-4e82-b28b-ebb72f8403e8.png)

**Comment by fabricio-godoi** [comment]
If anyone is still having problem with it, I found another workaround that should help a better user flow.

Before calling the `startActivityForSignInWithProvider`, use a WebView or directly call open the browser with the provider sign-in page. After the user authenticates in the provider, redirect it back to the application and then call the `startActivityForSignInWithProvider`. Since the user is already authenticated, it should get the authentication right away, without asking any more steps. Sure that it may have some more steps depending on the users steps (like not wanting to be asked the password or not again), but it is better than being unable to login.

I managed to perform a complete login with Microsoft using the following procedure:

1- Open web browser with the provider URL:
`https://login.microsoftonline.com/{{TENANT_ID}}/oauth2/v2.0/authorize?response_type=code&tenant={{TENANT_ID}}&client_id={{CLIENT_ID}}&redirect_uri={{LOGIN_REDIRECT}}&scope=profile%20email%20openid%20User.Read`

2- After the user has logged-in, the provider will redirect to the {{LOGIN_REDIRECT}} which in my case was a Dynamic Link for my application Login Screen;

3- Once I detect that WebView or Browser was closed, the application will continue calling the `startActivityForSignInWithProvider`, in which (in most cases) will perform without any user input;

Since both methods should open the user preferred Web Browser, the session storage in the browser will be used to proceed with the login.

Some notes:

 - You must configure your application to receive and manage properly the dynamic links;
 - This procedure may not work in iOS, since it is trickier to redirect the App to itself when running in some WebView (I was unable to do that yet);


Best regards!
